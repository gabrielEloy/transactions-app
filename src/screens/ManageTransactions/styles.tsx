import {StyleSheet} from 'react-native';
import {COLORS} from 'src/themes/main';

const styles = StyleSheet.create({
  header: {
    flex: 3,
    justifyContent: 'center',
    alignItems: 'center',
  },
  transactions: {
    flex: 5,
  },
  descriptionText: {
    textAlign: 'center',
    fontFamily: 'Exo-Medium',
    fontSize: 16,
    color: COLORS.gray,
  },
  totalText: {
    fontFamily: 'Exo-Light',
    fontSize: 42,
    color: COLORS.light,
    marginBottom: 10,
  },
  buttonText: {
    fontFamily: 'Exo-Medium',
    color: COLORS.light,
    fontSize: 18,
  },
});

export default styles;
