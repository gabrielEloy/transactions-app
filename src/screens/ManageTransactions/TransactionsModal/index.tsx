import React, {useState, useContext} from 'react';
import {View, Text, Modal, TextInput, Alert} from 'react-native';
import SwitchSelector from 'react-native-switch-selector';

import {COLORS} from 'src/themes/main';
import Button from 'src/components/Button';
import Divider from 'src/components/Divider';
import TransactionsContext from 'src/contexts/TransactionsContext';
import CurrencyInput from 'src/components/CurrencyInput';
import {getValueAndCurrency} from 'src/helpers/currency';
import styles from './styles';

interface Props {
  isModalVisible: boolean;
}

const TransactionsModal = (props: Props) => {
  const [description, setDescription] = useState('');
  const [value, setValue] = useState('');
  const [debit, setIsDebit] = useState(false);

  const context = useContext(TransactionsContext);

  const resetState = () => {
    setValue('');
    setDescription('');
    setIsDebit(false);
  };

  const handleSwitchSelect = (switchValue: string) => {
    setIsDebit(switchValue === 'Debit');
  };
  const handleValue = (val: string) => {
    setValue(val);
  };
  const handleDescription = (descValue: string) => setDescription(descValue);

  const checkFilledFields = () => {
    let errors = '';

    if (!description) {
      errors += '- missing description \n';
    }
    if (!value) {
      errors += '- missing value \n';
    }

    return errors;
  };

  const handleAddTransaction = () => {
    const errors = checkFilledFields();

    if (errors) {
      Alert.alert('Missing fields', errors);
      return;
    }

    context.addTransaction({
      description,
      value: getValueAndCurrency(value),
      debit,
    });
    context.setModalVisible(false);
    resetState();
  };

  const cancelAddTransaction = () => {
    resetState();
    context.setModalVisible(false);
  };

  const options = [
    {label: 'Credit', value: 'Credit', activeColor: COLORS.primary},
    {label: 'Debit', value: 'Debit', activeColor: COLORS.primary},
  ];

  return (
    <Modal transparent animationType="fade" visible={props.isModalVisible}>
      <View style={styles.modalStyle}>
        <View style={styles.modalAction}>
          <View style={styles.modalHeader}>
            <Text style={styles.headerText}>New Transaction</Text>
            <Divider />
          </View>
          <View style={styles.form}>
            <View>
              <Text style={styles.text}>Type</Text>
              <SwitchSelector
                initial={0}
                borderRadius={5}
                animationDuration={200}
                style={styles.switchStyle}
                options={options}
                onPress={handleSwitchSelect}
                textStyle={{...styles.text, ...styles.blackText}}
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.text}>Description</Text>
              <TextInput
                testID="descriptionInput"
                value={description}
                onChangeText={handleDescription}
                style={styles.input}
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.text}>Value</Text>
              <CurrencyInput
                testID="valueInput"
                value={value}
                onChangeText={handleValue}
                style={styles.input}
              />
            </View>
          </View>
          <View style={styles.footer}>
            <Button
              testID="cancelButton"
              onPress={cancelAddTransaction}
              removeBackground>
              <Text style={styles.text}>Cancel</Text>
            </Button>
            <Button
              testID="addTransactionButton"
              onPress={handleAddTransaction}>
              <Text style={styles.text}>Add</Text>
            </Button>
          </View>
        </View>
      </View>
    </Modal>
  );
};

export default TransactionsModal;
