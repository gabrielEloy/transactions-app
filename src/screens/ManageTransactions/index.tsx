import React, {useState, useEffect} from 'react';
import {View, Text} from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import moment from 'moment';

import Button from 'src/components/Button';
import TransactionsList from 'src/components/TransactionsList';
import {getCurrencyString} from 'src/helpers/currency';
import TransactionsContext from 'src/contexts/TransactionsContext';
import ITransactions from 'src/interfaces/ITransaction';
import IFormTransaction from 'src/interfaces/IFormTransaction';
import {sumValue} from 'src/helpers/initialState';
import TransactionsModal from './TransactionsModal';
import styles from './styles';

const ManageTransactions = () => {
  const [isModalVisible, setModalVisible] = useState(false);
  const [transactions, setTransactions] = useState<ITransactions[]>([]);
  const [total, setTotal] = useState(0);

  const getTotalValue = (evalutatedTransactions: ITransactions[]) =>
    setTotal(sumValue(evalutatedTransactions));

  const handleInitialState = async () => {
    const savedTransactions = await AsyncStorage.getItem('transactions');
    if (savedTransactions) {
      const parsedTransactions = JSON.parse(savedTransactions);

      setTransactions(parsedTransactions);
      getTotalValue(parsedTransactions);
    }
  };

  useEffect(() => {
    handleInitialState();
  }, []);

  //fazer isso virar lÃ³gica interna do updateTotal
  
  const updateTotal = (transaction: ITransactions, add: boolean) => {
    const transactionValue = transaction.debit
      ? -transaction.value.value
      : transaction.value.value;

    const newTotal = add ? transactionValue + total : -transactionValue + total;

    setTotal(newTotal);
  };


  //trocar addTransaction por um hook

  const addTransaction = async (transactionForm: IFormTransaction) => {
    const date = moment().toISOString();
    const id = `${date}-${transactionForm.description}`;
    const addedTransaction = {...transactionForm, date, id};
    const updatedTransactions = [addedTransaction, ...transactions];

    setTransactions(updatedTransactions);
    updateTotal(addedTransaction, true);

    await AsyncStorage.setItem(
      'transactions',
      JSON.stringify(updatedTransactions),
    );
  };

  const removeTransaction = async (removedTransaction: ITransactions) => {
    const filteredTransaction = transactions.filter(
      (transaction) => transaction.id !== removedTransaction.id,
    );

    setTransactions(filteredTransaction);
    updateTotal(removedTransaction, false);

    await AsyncStorage.setItem(
      'transactions',
      JSON.stringify(filteredTransaction),
    );
  };
  
  return (
    <TransactionsContext.Provider
      value={{addTransaction, setModalVisible, removeTransaction}}>
      <TransactionsModal isModalVisible={isModalVisible} />
      <View style={styles.header}>
        <Text style={styles.descriptionText}>Total</Text>
        <Text style={styles.totalText}>
          {getCurrencyString({value: total, currency: 'R$'})}
        </Text>
        <Button
          testID="rootAddTransactionButton"
          onPress={() => setModalVisible(!isModalVisible)}>
          <Text style={styles.buttonText}>Add transaction</Text>
        </Button>
      </View>
      <View style={styles.transactions}>
        <Text style={styles.descriptionText}>Transactions</Text>
        <TransactionsList
          transactions={transactions}
          onSwipe={removeTransaction}
        />
      </View>
    </TransactionsContext.Provider>
  );
};

export default ManageTransactions;
