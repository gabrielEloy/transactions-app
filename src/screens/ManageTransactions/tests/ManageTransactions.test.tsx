import React from 'react';
import {render, fireEvent} from 'react-native-testing-library';
import ManageTransactions from '../index';

describe('test <ManageTransactions />', () => {
  test('should successfully render component', () => {
    const tree = render(<ManageTransactions />);
    expect(tree).toMatchSnapshot();
  });

  test('should render modal but keep it invisible', async () => {
    const {getByTestId} = await render(<ManageTransactions />);

    const modal = await getByTestId('modal');

    expect(modal.props.visible).toBeFalsy();
  });

  test('should change modal visibility on click the add transaction button', () => {
    const {getByTestId} = render(<ManageTransactions />);

    const addTransactionButton = getByTestId('rootAddTransactionButton');

    fireEvent.press(addTransactionButton);
    const modal = getByTestId('modal');

    expect(modal.props.visible).toBeTruthy();
  });

  test('should succesfully add transaction', () => {
    const {getByTestId} = render(<ManageTransactions />);

    const transactionInitialListDataLength = getByTestId('transactionList')
      .props.data.length;

    const addTransactionButton = getByTestId('rootAddTransactionButton');
    const modalAddTransactionButton = getByTestId('addTransactionButton');
    const descriptionInput = getByTestId('descriptionInput');
    const valueInput = getByTestId('valueInput');

    fireEvent.press(addTransactionButton);

    fireEvent.changeText(descriptionInput, 'received money');
    fireEvent.changeText(valueInput, '22,50');

    fireEvent.press(modalAddTransactionButton);

    const transactionFinalListDataLength = getByTestId('transactionList').props
      .data.length;
    expect(transactionFinalListDataLength).toBeGreaterThan(
      transactionInitialListDataLength,
    );
  });

  test('should succesfully add transaction', () => {
    const {getByTestId} = render(<ManageTransactions />);

    const transactionInitialListDataLength = getByTestId('transactionList')
      .props.data.length;

    const addTransactionButton = getByTestId('rootAddTransactionButton');
    const modalAddTransactionButton = getByTestId('addTransactionButton');
    const descriptionInput = getByTestId('descriptionInput');
    const valueInput = getByTestId('valueInput');

    fireEvent.press(addTransactionButton);

    fireEvent.changeText(descriptionInput, 'received money');
    fireEvent.changeText(valueInput, '22,50');

    fireEvent.press(modalAddTransactionButton);

    const transactionFinalListDataLength = getByTestId('transactionList').props
      .data.length;
    expect(transactionFinalListDataLength).toBeGreaterThan(
      transactionInitialListDataLength,
    );
  });
});
