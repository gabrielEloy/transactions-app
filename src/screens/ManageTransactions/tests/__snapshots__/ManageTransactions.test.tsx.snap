// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <ManageTransactions /> should successfully render component 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0,0,0,.3)",
          "borderRadius": 10,
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#2D2D38",
            "borderRadius": 20,
            "height": 450,
            "overflow": "hidden",
            "width": "80%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "paddingVertical": 20,
              "position": "relative",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FBFBFB",
                "fontSize": 18,
              }
            }
          >
            New Transaction
          </Text>
          <View
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                },
                Object {
                  "backgroundColor": "#7A7A80",
                  "height": 1,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 2,
              "padding": 20,
            }
          }
        >
          <View>
            <Text
              style={
                Object {
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
            >
              Type
            </Text>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "borderRadius": 0,
                    "marginTop": 20,
                  },
                ]
              }
            >
              <View
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <View
                  onLayout={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "height": 40,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "borderColor": "#C9C9C9",
                        "borderRadius": 5,
                        "borderWidth": 0,
                        "flex": 1,
                        "flexDirection": "row",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "opacity": 1,
                        }
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                              "color": "#FFFFFF",
                              "fontSize": 14,
                              "fontWeight": "normal",
                              "textAlign": "center",
                            },
                            Object {},
                          ]
                        }
                      >
                        Credit
                      </Text>
                    </View>
                    <View
                      accessible={true}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "opacity": 1,
                        }
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                              "color": "#000000",
                              "fontSize": 14,
                              "fontWeight": "normal",
                              "textAlign": "center",
                            },
                            Object {
                              "color": "black",
                              "fontFamily": "Exo-Medium",
                            },
                          ]
                        }
                      >
                        Debit
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
            >
              Description
            </Text>
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              rejectResponderTermination={true}
              style={
                Object {
                  "borderBottomColor": "#7A7A80",
                  "borderBottomWidth": 1,
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
              testID="descriptionInput"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            style={
              Object {
                "marginTop": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
            >
              Value
            </Text>
            <TextInput
              allowFontScaling={true}
              keyboardType="numeric"
              onChangeText={[Function]}
              options={
                Object {
                  "delimiter": ".",
                  "precision": 2,
                  "separator": ",",
                  "suffixUnit": "",
                  "unit": "R$ ",
                }
              }
              rejectResponderTermination={true}
              style={
                Object {
                  "borderBottomColor": "#7A7A80",
                  "borderBottomWidth": 1,
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
              testID="valueInput"
              type="money"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "marginBottom": 20,
              "marginRight": 25,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,0)",
                "borderRadius": 5,
                "opacity": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 15,
              }
            }
            testID="cancelButton"
          >
            <Text
              style={
                Object {
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
            >
              Cancel
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#903EE3",
                "borderRadius": 5,
                "opacity": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 15,
              }
            }
            testID="addTransactionButton"
          >
            <Text
              style={
                Object {
                  "color": "#FBFBFB",
                  "fontFamily": "Exo-Medium",
                }
              }
            >
              Add
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>,
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#7A7A80",
          "fontFamily": "Exo-Medium",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      Total
    </Text>
    <Text
      style={
        Object {
          "color": "#FBFBFB",
          "fontFamily": "Exo-Light",
          "fontSize": 42,
          "marginBottom": 10,
        }
      }
    >
      R$ 0,00
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#903EE3",
          "borderRadius": 5,
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
      testID="rootAddTransactionButton"
    >
      <Text
        style={
          Object {
            "color": "#FBFBFB",
            "fontFamily": "Exo-Medium",
            "fontSize": 18,
          }
        }
      >
        Add transaction
      </Text>
    </View>
  </View>,
  <View
    style={
      Object {
        "flex": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#7A7A80",
          "fontFamily": "Exo-Medium",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      Transactions
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="transactionList"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>,
]
`;
